---
title: "Diaptera wings classification using Topological Data Analysis"
author:
  - name: Guilherme Vituri F. Pinto
    orcid: 0000-0002-7813-8777
    corresponding: true
    email: vituri.vituri@gmail.com
    roles:
      - Investigation
      - Project administration
      - Software
      - Visualization
    affiliations:
      - Universidade Estadual Paulista
  - name: Sergio T. Ura
    orcid: 0000-0002-5414-6094
    corresponding: true
    email: sergio.ura@gmail.com
    roles:
      - Investigation
      - Project administration
      - Software
      - Visualization
    affiliations:
      - Universidade Estadual Paulista
  - name: Northon C. Leme Penteado
    orcid: 0000-0002-9466-9906
    corresponding: true
    email: northon.canevari@ufca.edu.br
    roles:
      - Investigation
      - Project administration
      - Software
      - Visualization
    affiliations:
      - Universidade Federal do Cariri
keywords:
  - Topological Data Analysis
  - Persistent homology
  - Diaptera classification
abstract: |
  We studied etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc 
plain-language-summary: |
  We studied etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc  etc etc etc 
key-points:
  - Images taken from etc.
  - Analysis resulted in etc.
  - Results were etc.
date: last-modified
bibliography: references.bib
citation:
  container-title: Earth and Space Science
number-sections: true
engine: julia
---

```{julia}
using TDAfly, TDAfly.Preprocessing, TDAfly.TDA, TDAfly.Analysis
using Images: mosaicview
using Plots: plot, display, heatmap, scatter
using PersistenceDiagrams
```

## Introduction

A taxonomy (or taxonomic classification) is a scheme of classification, especially a hierarchical classification, in which things are organized into groups or types. Correct insect identification is critically important in
entomological research, pest control, and insect-resource utilization. However, to identify an unknown species using traditional methods is very time consuming due to the vast number of insect species that must be identified
and limited taxonomist resource @gaston1 and @gaston2.

For insect groups, wing morphology is an important character for species identification. 
 
This problem has been recognized since the end of the last century, and computer-based recognition systems have been developed to identify insect species automatically @weeks. For example, the “Digital Automated Identification System (DAISY)” @weeks2, @weeks3 and @watson was developed to automatically identify ichneumonid wasps on the basis of their wing features @weeks, and this program was also used to identify biting midges and moths @watson. The “Species Identified Automatically (SPIDA)” system was originally designed to distinguish between 121 species of Australian spiders @platnick and has been successfully tested for identifying wasps and bees using their wing characteristics @weeks3. In addition, the “Automatic Bee Identification System (ABIS)” was developed to identify bees by analyzing images of wing veins @francoi.

All of these automatic identification systems involve two key processes: extracting and digitizing the taxonomic features and constructing a classifier model for identification.

First, a recognition feature must be selected. The feature must be not only stable and distinguishable but also extractable and quantifiable by computer. The feature “color” was successfully used to identify moths and butterflies @qing, but its accuracy was limited by specimens fading with time or color shifts under different illumination. 

In recent decades, software has been developed to digitize morphological characteristics and conduct consequent statistical analysis, such as principal component analysis and canonical correlation analysis, to establish the phylogenic relations of species. Examples of this software include TPSDig @rohlf, DrawWing @tofilski, @mutanen, @mutanen2 and SHAPE @iwata and @yoshioka, which are based on modern digitalizing methods, such as ‘Landmark’ and ‘Semilandmark’ @rohlf and @tofilski. 

Most adult insects have wings that are frequently used in taxonomy. However, the detailed characteristics of veins and spots on wings require recognition and measurement by well-trained experts; thus, these features are difficult for general users to use. 

The objective of this work is to use methods from TDA (Topological Data Analysis) to obtain a classification based on images of fly wings. Specifically, based on the venation extracted from those images.

## Methods

The anlysis start to load the wings images from a directory and for each image apply a gaussian blur, crop and to make it binary image and standardize the size with 150 pixels of height. The blurring step is necessary to disregard small holes in the figure and keep it connected.

```{julia}
paths = readdir("C:/Users/north/Documents/tda-fly/images/processed", join=true)
species = basename.(paths) .|> (x -> replace(x, ".png" => ""))
individuals = map(species) do specie
  s = split(specie, " ")
  s[1][1] * "-" * s[2]
end
wings = load_wing.(paths, blur = 1.3)
Xs = map(image_to_r2, wings);
```

```{julia}
mosaicview(wings, ncol = 4, fillvalue=1)
```

### Vietoris-Rips filtration

We select 500 points from each image using a farthest point sample method ensuring a representative selection for wing venation

```{julia}
samples = map(Xs) do X
  ids = farthest_points_sample(X, 500)
  X[ids]
end;
```

and create an empty dictionary to store all computations

```{julia}
simple_rips_dc = Dict();
```

We then calculate its persistence diagrams using the Vietoris-Rips filtration.

```{julia}
# get only the 1-dimensional PD
simple_rips_dc["PD"] = rips_pd.(samples, cutoff = 5, threshold = 200) .|> last;
```

We create the 1-dimensional persistence image for each persistence diagram using 10x10 matrices.

```{julia}
PI = PersistenceImage(simple_rips_dc["PD"], size = (10, 10))

simple_rips_dc["PI"] = PI.(simple_rips_dc["PD"]);
```

#### Examples

Below are some examples of 1-dimensional barcodes, its persistence image and the original wing that generated it. Note: we are plotting the barcode using the birth and persistence.

```{julia}
# plot some images to see the barcodes
map([1, 4, 8, 10, 15]) do i
  p = plot_wing_with_pd(simple_rips_dc["PD"][i], simple_rips_dc["PI"][i], samples[i], species[i])
  display(p)
end;
```


We now calculate the Euclidean distance between each persistence image (seen as a vector of $\mathbb{R}^{10x10}$) and plot its heatmap.

```{julia}
simple_rips_dc["Distance_matrix"] = pairwise_distance(simple_rips_dc["PI"]);
```

```{julia}
plot_heatmap(
  simple_rips_dc["Distance_matrix"], 
  individuals, 
  "Distance matrix for Vietoris-Rips barcodes"
)
```

### Persistence Homology Transform

From the 500 selected points we use another method of complex filtration based on a line direction to study and compare the generated persistence diagrams. 


We start with the point (0, 0). Its filtration is the following

```{julia}
A = wings[10] |> image_to_array;
f = dist_to_point(0, 0)
Af = modify_array(A, f)
heatmap(Af)
```

with corresponding sublevel barcode as

```{julia}
point_pds = cubical_pd(Af, cutoff = 0.05)
plot_pd(point_pds)
```

or, with persistence in the y-axis:

```{julia}
plot_pd(point_pds, persistence = true)
```

Let's see step-by-step of this filtration:

```{julia}
for tr ∈ reverse([0:0.1:1;])
  X = findall_ids(>(tr), Af)
  title = "threshold: $tr"
  p = scatter(first.(X), last.(X), title = title)
  display(p)
end
```

Due to noise, some connected components are born in 0.2 and die only at 0. But the loops seems alright.

Let's apply this idea to all images!

```{julia}
point_rips_dc = Dict()
f = dist_to_point(0, 0)

point_rips_dc["PD"] = map(wings) do wing
  A = wing |> image_to_array;
  Af = modify_array(A, f)
  point_pds = cubical_pd(Af, cutoff = 0.05)[2]
end;

PI = PersistenceImage(point_rips_dc["PD"], size = (10, 10))
point_rips_dc["PI"] = PI.(point_rips_dc["PD"]);
```

```{julia}
final_dc = Dict();

dics = [simple_rips_dc, point_rips_dc]

final_dc["PI"] = simple_rips_dc["PI"] + point_rips_dc["PI"];
```

```{julia}
final_dc["Distance_matrix"] = pairwise_distance(final_dc["PI"]);
```

```{julia}
plot_heatmap(
  final_dc["Distance_matrix"], 
  individuals, 
  "Distance matrix for Vietoris-Rips + point-based barcodes"
)
```

What happened?! Some wing has a big distance from everything else!

```{julia}
Xs = point_rips_dc["PI"]
id = findfirst(==("c-39"), individuals)
# id = 9
X = Xs[id]
heatmap(X)
```

```{julia}
wings[id]
```

```{julia}
heatmap(wings[id])
```

The reason is that it has one small cycle right at the beggining of the filtration...

## Draft.....

```{julia}
#| include: false 
ps = map(reverse([0:0.1:0.9;])) do tr
  X = findall_ids(>(tr), Af)
  title = "threshold: $tr"
  p = scatter(first.(X), last.(X), title = title)
  # display(p)
end

plot(ps...)
```

### References

::: {#refs}
:::